lib/RunTests.js

```diff
@@ -0,0 +1,274 @@
+      // within 60 ms on Windows, and faster on Mac and Linux. So 100 ms sounds
+      // like a reasonable number – not too short, not too long of a wait.
+      if (queue.length > 0) {
+        await delay(100);
+        // Re-print the message in case the queue has become longer while waiting.
+        clearConsole(report);
+        infoLog(report, watcherEventMessage(queue));
+      }
+
+      queue = [];
+
+      // All operations down to `Generate.generateElmJson(project)` could
+      // potentially be avoided depending on what changed (checking what’s
+      // inside `queue`). But at the time of this writing all of them are fast
+      // (less than 20 ms, often less than 1 ms) so it’s not worth bothering.
+
+      // Files may be changed, added or removed so always re-create project info
+      // from disk to stay fresh.
+      const project = Project.init(projectRootDir, packageInfo.version);
+      Project.validateTestsSourceDirs(project);
+
+      const testFilePaths = FindTests.resolveGlobs(
+        testFileGlobs.length === 0 ? [project.testsDir] : testFileGlobs,
+        project.rootDir
+      );
+
+      if (testFilePaths.length === 0) {
+        throw new Error(
+          FindTests.noFilesFoundError(project.rootDir, testFileGlobs)
+        );
+      }
+
+      if (watcher !== undefined) {
+        const diff = diffArrays(watchedPaths, project.testsSourceDirs);
+        watchedPaths = project.testsSourceDirs;
+        watcher.add(diff.added);
+        watcher.unwatch(diff.removed);
+      }
+
+      Generate.generateElmJson(project);
+
+      const testModules = await FindTests.findTests(testFilePaths, project);
+
+      const mainFile = Generate.generateMainModule(
+        fuzz,
+        seed,
+        report,
+        testFileGlobs,
+        testFilePaths,
+        testModules,
+        project.generatedCodeDir,
+        processes
+      );
+
+      const dest = path.join(project.generatedCodeDir, 'elmTestOutput.js');
+
+      runsExecuted++;
+      const pipeFilename = getPipeFilename(runsExecuted);
+
+      await Compile.compile(
+        project.generatedCodeDir,
+        mainFile,
+        dest,
+        pathToElmBinary,
+        report
+      );
+
+      Generate.prepareCompiledJsFile(pipeFilename, dest);
+
+      return await Supervisor.run(
+        packageInfo.version,
+        pipeFilename,
+        report,
+        processes,
+        dest,
+        watch
+      );
+    } catch (err) {
+      console.error(err.message);
+      return 1;
+    }
+  }
+
+  if (watch) {
+    clearConsole(report);
+    infoLog(report, 'Running in watch mode');
+
+    const onRunFinish = () => {
+      if (queue.length > 0) {
+        clearConsole(report);
+        infoLog(report, watcherEventMessage(queue));
+        currentRun = run().then(onRunFinish);
+      } else {
+        infoLog(report, chalk.blue('Watching for changes...'));
+        currentRun = undefined;
```

Comment:
minor: if not planning to take advantage of invoking default arg handling semantics passing `currentRun` anywhere, then recommend avoiding `undefined` assignment as it is somewhat uniquely quirky to JS. i.e. prefer `null`

Reply:
I disagree – I find `null` much quirkier than `undefined`

Reply:
Heh, ok